projects/transformers/trainer_mixins/one_cycle_lr.py

```diff
@@ -0,0 +1,85 @@
+# ----------------------------------------------------------------------
+# Numenta Platform for Intelligent Computing (NuPIC)
+# Copyright (C) 2021, Numenta, Inc.  Unless you have an agreement
+# with Numenta, Inc., for a separate license for this software code, the
+# following terms and conditions apply:
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU Affero Public License version 3 as
+# published by the Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+# See the GNU Affero Public License for more details.
+#
+# You should have received a copy of the GNU Affero Public License
+# along with this program.  If not, see http://www.gnu.org/licenses.
+#
+# http://numenta.org/licenses/
+# ----------------------------------------------------------------------
+
+from torch.optim.lr_scheduler import OneCycleLR
+
+
+class OneCycleLRMixin:
+    """
+    Mixin to HF Trainer for using the `OneCycleLR`_. See documentation for argument
+    details.
+
+    .. _OneCycleLR:
+        https://pytorch.org/docs/stable/optim.html#torch.optim.lr_scheduler.OneCycleLR
+    """
+
+    def __init__(
+        self,
+        max_lr=1e-2,
+        pct_start=0.3,
+        anneal_strategy="linear",
+        cycle_momentum=True,
+        base_momentum=0.85,
+        max_momentum=0.95,
+        div_factor=25,
+        final_div_factor=1e4,
+        last_epoch=-1,
+        **kwargs,
+    ):
+        super().__init__(**kwargs)
+
+        self.max_lr = max_lr
+        self.pct_start = pct_start
+        self.anneal_strategy = anneal_strategy
+        self.cycle_momentum = cycle_momentum
+        self.base_momentum = base_momentum
+        self.max_momentum = max_momentum
+        self.div_factor = div_factor
+        self.final_div_factor = final_div_factor
+        self.last_epoch = last_epoch
+
+    def create_optimizer_and_scheduler(self, num_training_steps: int):
+        """
+        Setup the optimizer and the learning rate scheduler. This overrides super
+        in a way that just customizes the lr scheduler while the optimizer remains the
+        default.
+        """
+
+        # Set lr scheduler to dummy variable so it's not created in the call to super.
+        self.lr_scheduler = ...
+
+        # Create just the optimizer.
+        super().create_optimizer_and_scheduler(num_training_steps)
+
+        # Now define the lr scheduler, given the optimizer.
+        self.lr_scheduler = OneCycleLR(
```

Comment:
Instead of setting lr_scheduler, calling create_optimizer_and_scheduler and then redefining lr_scheduler, why not override `Trainer.create_scheduler` instead?

Reply:
I don't believe there is such a class method, at least in version 4.4.2. There is `get_scheduler`, but that belongs to `transformers.optimization`.

Reply:
Interesting. There is such a function in 4.5
![image](https://user-images.githubusercontent.com/641561/113644071-b392a800-9638-11eb-9850-d38269e107e6.png)

Reply:
I'm now on the cautious side for upgrading after seeing how quickly their repo changes. Last I tried 4.5.0, there was the cloudpickle issue when attempting a hyperparameter search. I might try it out in the next week or so, and we can discuss upgrading.

Reply:
Related: We should have tests for our mixins. As we upgrade the transformers version, things are likely to break if they keep refactoring the code as much as they do. I'll try and add some for onecycle lr, the lr range tests, and rigl that minimally invoke the trainer somehow. Hopefully, this could provide a template for your kd tests.

Reply:
Don't 100% agree with you. Yes, things are going to break, but we can also just fix them as they break. Tests are wonderful for products that require reliability, but they also take time, and in the past we've lost a significant amount of time building elaborate test suites for things we end up not using a few months later.

Reply:
I have some tests in my task list, for example, but the priority assigned to them is such that there is a small chance they make to the top of that list soon, unless our team grows fast or our deadlines are postponed.